packages/engine/src/simulation/package/state/packages/behavior_execution/package.py

```diff
@@ -58,88 +59,108 @@ def load_behaviors(behavior_descs):
+    #       this try/except inside the loop.
+    try:
+        position = agent_state.position
+        if position is not None:  # TODO: OPTIM Is position nullable?
+            while len(position) < 3:
+                position.append(0.0)
+    except AttributeError:
+        pass
+
+    # TODO: OPTIM Same todos as for position
+    try:
+        direction = agent_state.direction
+        if direction is not None:
+            while len(direction) < 3:
+                direction.append(0.0)
+    except AttributeError:
+        pass
+
+
+# For each agent in the given group, execute behaviors and postprocess state,
+# starting after the last behavior already executed (during this step / more generally
+# behavior execution package call) and stopping when all behaviors are executed or
+# the next behavior is in a different language (i.e. not Python).
+def run_task(experiment, _sim, _task_message, group_state, group_context):
     next_lang = None
     agent_state = None
     agent_context = None
-    
-    for i_agent in range(group_state.n_agents()):
-    
-        # Reuse `agent_state` object.
-        agent_state = group_state.get_agent(i_agent, agent_state)
 
-        behavior_ids = agent_state.__behaviors
-        i_behavior = agent_state.__i_behavior # Need `i_behavior` outside loop scope.
-        while i_behavior < len(behavior_ids):
-            agent_state.__i_behavior = i_behavior
+    # TODO: Propagate field specs to runners and use in state and context objects
+    behavior_ids_field_key = '_PRIVATE_14_behavior_ids'
 
-            behavior = experiment.behaviors[hash_behavior_id(behavior_ids[i_behavior])]
-            if behavior.language != "js":
-                next_lang = behavior.language  # Multiple assignments are fine.
+    for i_agent in range(group_state.n_agents()):
+        # Reuse `agent_state` and `agent_context` objects.
+        agent_state = group_state.get_agent(i_agent, agent_state)
+        agent_context = group_context.get_agent(i_agent, agent_context)
+
+        # ids of behaviors of this agent
+        behavior_ids = getattr(agent_state, behavior_ids_field_key)
+
+        # `behavior_index` is the index of the first behavior that
+        # hasn't been executed yet (during this step / package call).
+        for i_behavior in range(int(agent_state.behavior_index), len(behavior_ids)):
+            agent_state.behavior_index = i_behavior  # Keep up to date for use in behaviors.
+
+            # Behavior ids are shallow-loaded as an optimization, so
+            # need to convert to a Python object here.
+            # TODO: OPTIM Use `.values` attribute after upgrading Arrow.
+            b_id = behavior_ids[i_behavior].as_py()
+            behavior = experiment['behaviors'][_hash_behavior_id(b_id[0], b_id[1])]
+            if behavior['language'] != "Python":
+                next_lang = behavior['language']  # Multiple assignments are fine.
```

Comment:
What does the comment mean?

Reply:
`next_lang` might be assigned to multiple times and overwritten, but that's ok, because we can just use the last value that was assigned to it.